Add handing of SMTP server error messages

dmitryalexeyev opened the pull request:
…mtp server or reports meaningless error messages if gets a.e. 5xx error codes.

This patch adds handling of curl errors if smtp server drops connection or doesn't reply correctly on rcpt command.
It also provides test for SMTP error codes 3xx, 4xx, 5xx and unexpected errors handling.

Fixes #13

Totktonada commented on smtp/smtpc.c:
```diff
@@ -438,14 +450,23 @@ smtpc_execute(struct smtpc_request *req, double timeout)
 		snprintf(error_msg, sizeof(error_msg), "SMTP error %i (os errno %li)", req->code, longval);
 		box_error_set(__FILE__, __LINE__, ER_UNKNOWN, error_msg);
 		++env->stat.failed_requests;
+		free(req->curl_error);
 		smtpc_request_delete(req);
```
It seems, we'll free `curl_error` twice in this case.

Totktonada commented on smtp/lib.c:
```diff
@@ -78,7 +78,7 @@ luaT_smtpc_request(lua_State *L)
 		return luaT_error(L);
 
-	double timeout = 365 * 24 * 3600;
+	double timeout = 10;
```
The default timeout is changed here for all usages, not just for tests.

dmitryalexeyev replied:
Yes, I know this one isn't very good. But 10 seconds timeout for RCPT command seems reasonable to me, something wrong must be happening if we hit it.
Not sure about 1 year timeout om RCPT command, we need to bail out sooner or later if something went wrong. And we won't get to DATA cmd until we pass RCPT, so we hang forever here.
Here comes second issue - 10 seconds timeout works fine on mac, but seems to be ignored on Linux.

Shall we distinguish test case and normal work? Shall we use 1 year timeout on RCPT? Doubt it, probably 10 to 30 seconds would be better

Totktonada replied:
Sure, infinite timeouts are generally discouraged. But it is not the reason to change it here: it is irrelevant to the issue. And we definitely should not change it silently: without any mention in the commit message (but better move to its own commit).

AFAIU, we should pass the timeout option to test the fix of the error message and something went wrong with this functionality on Linux. If it is so, it looks as a bug, which blocks testing of this change. I suggest to file an issue about this with a small reproducer and tarantool/libcurl/module versions (plus platform: distribution and distro version).

dmitryalexeyev replied:
I see; my bad. I missed the point I can control timeout from lua code directly.

> If it is so, it looks as a bug, which blocks testing of this change. I suggest to file an issue about this with a small reproducer and tarantool/libcurl/module versions (plus platform: distribution and distro version).

Sure, will do.

dmitryalexeyev replied:
The issue is smtp_h spamming 502 Not implemented every time it receives empty reply, we just needed return statement after it. Not related to curl timeout in any way.
Fixed.

Totktonada commented on smtp/smtpc.c:
```diff
@@ -250,8 +250,17 @@ smtpc_request_new(struct smtpc_env *env, const char *url, const char *from)
 		return NULL;
 	}
+
+	req->curl_error = (char *) malloc(CURL_ERROR_SIZE);
```
Why don't allocate it as part of the request structure?

Totktonada commented:
(The comments above are just after slight glance, I didn't looked into the patch decently.)

It seems, it supersedes PR #45. Please, use force-pushes rather than creating a new PR for each new version of a patchset. This way we'll have the discussion in one place.

dmitryalexeyev commented:
The problem is I deleted the repository, I originally created pr #45 from :(

dmitryalexeyev committed:
According to #13 smtpc hangs if receives unexpected error code from smtp server or reports meaningless error messages if gets a.e. 5xx error codes.
This patch adds handling of curl errors if smtp server drops connection or doesn't reply correctly on rcpt command.
It also provides test for SMTP error codes 3xx, 4xx, 5xx and unexpected errors handling.

Fixes #13
